llvm/tools/ir2builder/ir2builder.cpp

```diff
@@ -0,0 +1,1702 @@
+  std::string getCallingConv(CallingConv::ID cc);
+  std::string getConstantRange(ConstantRange &cr);
+
+  std::string getVal(const Value *op);
+  std::string getConst(const Constant *op);
+  std::string getType(const Type *t);
+  std::string getInlineAsm(const InlineAsm *op);
+  std::string getMetadata(const Metadata *op);
+
+public:
+  IR2Builder()
+      : llvmPrefix(ScopePrefix), builder(BuilderName), ctx(ContextName),
+        modName(ModuleName) {}
+
+  /**
+   * Calls convert for all the functions in passed in module
+   * @param M Module to call convert over
+   * @param OS Stream to which output the builder calls
+   */
+  void convert(Module &M, raw_ostream &OS);
+
+  /**
+   * Converts a function into IRBuilder API calls
+   * @param F Function to convert
+   * @param OS Stream to which output the builder calls
+   */
+  void convert(Function &F, raw_ostream &OS);
+
+  /**
+   * Converts an instruction into IRBuilder API calls
+   * @param I Instruction to convert
+   * @param OS Stream to which output the builder calls
+   * @note Unsupported instructions or their operands should result
+   *       in a TODO comment.
+   */
+  void convert(const Instruction *I, raw_ostream &OS);
+};
+
+std::string IR2Builder::getNextVar() {
+  return "v0" + std::to_string(varI++);
+}
+
+static std::string to_str(bool b) { return b ? "true" : "false"; }
+
+static std::string escape(std::string str) {
+  std::stringstream ss;
+  for (unsigned char C : str) {
+    if (C == '\\')
+      ss << '\\' << C;
+    else if (isPrint(C) && C != '"')
+      ss << C;
+    else
+      ss << '\\' << hexdigit(C >> 4) << hexdigit(C & 0x0F);
+  }
+  return ss.str();
+}
+
+static std::string sanitize(std::string s) {
+  std::stringstream ss;
+  for (size_t i = 0; i < s.size(); ++i) {
+    if (!std::isalnum(s[i])) {
+      ss << "_" << static_cast<unsigned>(s[i]) << "_";
+    } else {
+      ss << s[i];
+    }
+  }
+  return ss.str();
+}
+
+std::string IR2Builder::getLinkage(llvm::GlobalValue &gv) {
+  std::string link = llvmPrefix + "GlobalValue::";
+
+  if (gv.hasExternalLinkage()) {
+    return link + "ExternalLinkage";
+  }
+  if (gv.hasAvailableExternallyLinkage()) {
+    return link + "AvailableExternallyLinkage";
+  }
+  if (gv.hasLinkOnceAnyLinkage()) {
+    return link + "LinkOnceAnyLinkage";
+  }
+  if (gv.hasLinkOnceODRLinkage()) {
+    return link + "LinkOnceODRLinkage";
+  }
+  if (gv.hasWeakAnyLinkage()) {
+    return link + "WeakAnyLinkage";
+  }
+  if (gv.hasWeakODRLinkage()) {
+    return link + "WeakODRLinkage";
+  }
+  if (gv.hasAppendingLinkage()) {
+    return link + "AppendingLinkage";
+  }
+  if (gv.hasInternalLinkage()) {
+    return link + "InternalLinkage";
+  }
+  if (gv.hasPrivateLinkage()) {
+    return link + "PrivateLinkage";
+  }
+  if (gv.hasExternalWeakLinkage()) {
+    return link + "ExternalWeakLinkage";
+  }
+  if (gv.hasCommonLinkage()) {
+    return link + "CommonLinkage";
+  }
+
+  return "/* Unknown linkage */";
+}
```

Comment:
To be consistent with your other rendering functions, you can take specifically a `Linkage` type instead of a `GlobalValue`. Then use LLVM's `GV.getLinkage()` as an argument to this routine. Something like this (I've called it `asStr` instead of `getLinkage`-- see earlier comments in the review-- this also avoids the having the same name as LLVM function):
```
  std::string IR2Builder::asStr(GlobalValue::LinkageTypes Linkage) {
  std::string link = llvmPrefix + "GlobalValue::";
  switch (Linkage) {
   case GlobalValue::WeakODRLinkage:
    return link + "WeakODRLinkage";
   case GlobalValue::LinkOnceODRLinkage:
    return link + "LinkOnceODRLinkage";
   case GlobalValue::AvailableExternallyLinkage:
    return link + "AvailableExternallyLinkage";
   case GlobalValue::WeakAnyLinkage:
    return link + "WeakAnyLinkage";
   case GlobalValue::LinkOnceAnyLinkage:
    return link + "LinkOnceAnyLinkage";
   case GlobalValue::CommonLinkage:
    return link + "CommonLinkage";
   case GlobalValue::ExternalWeakLinkage:
    return link + "ExternalWeakLinkage";
   case GlobalValue::ExternalLinkage:
    return link + "ExternalLinkage";
   case GlobalValue::AppendingLinkage:
    return link + "AppendingLinkage";
   case GlobalValue::InternalLinkage:
    return link + "InternalLinkage";
   case GlobalValue::PrivateLinkage:
    return link + "PrivateLinkage";
   default:
    return "/* Unknown linkage */";
   }
}
```